File: easy/0589_n_ary_tree_preorder_traversal/Solution.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
import nodes.Node;

class Solution {
  public List<Integer> preorder(Node root) {
    final List<Integer> vals = new ArrayList<>();

    if (root == null) {
      return vals;
    }

    final Deque<Node> stack = new ArrayDeque<>();
    stack.push(root);

    while (!stack.isEmpty()) {
      final Node node = stack.pop();

      vals.add(node.val);

      for (final Node child : node.children.reversed()) {
        stack.push(child);
      }
    }

    return vals;
  }
}

File: easy/0589_n_ary_tree_preorder_traversal/solution.py
from collections import deque
from typing import List

from utils.python.nodes import Node


class Solution:
    def preorder(self, root: "Node") -> List[int]:
        if not root:
            return []

        vals = []
        dq = deque([root])

        while dq:
            node = dq.popleft()

            vals.append(node.val)

            for child in reversed(node.children):
                dq.appendleft(child)

        return vals
